scripts/fine_tune.py

```diff
@@ -0,0 +1,142 @@
+    data_module = PathContextDataModule(config, vocabulary)
+    return model, data_module
+
+
+def load_typed_code2seq(
+    checkpoint_path: str, config: DictConfig, vocabulary: Vocabulary
+) -> Tuple[TypedCode2Seq, TypedPathContextDataModule]:
+    model = TypedCode2Seq.load_from_checkpoint(checkpoint_path=checkpoint_path)
+    data_module = TypedPathContextDataModule(config, vocabulary)
+    return model, data_module
+
+
+KNOWN_MODELS = {"code2seq": load_code2seq, "code2class": load_code2class, "typed-code2seq": load_typed_code2seq}
+
+
+def train_and_test(dataset_path: str, model_path: str, project_name: str, fold_idx: int) -> str:
+    """Trains model and return a path to best checkpoint"""
+
+    checkpoint = torch.load(model_path, map_location=torch.device("cpu"))
+    config = checkpoint["hyper_parameters"]["config"]
+    vocabulary = checkpoint["hyper_parameters"]["vocabulary"]
+    config.data_folder = dataset_path
+
+    if config.name not in KNOWN_MODELS:
+        print(f"Unknown model {config.name}, try one of {' '.join(KNOWN_MODELS.keys())}")
+        return ""
+
+    model, data_module = KNOWN_MODELS[config.name](model_path, config, vocabulary)
+
+    # define model checkpoint callback
+    checkpoint_callback = ModelCheckpoint(
+        dirpath=os.path.join("models", "fine_tuned", project_name, str(fold_idx)),
+        period=config.save_every_epoch,
+        save_top_k=-1,
+    )
+
+    # define other callbacks
+    early_stopping_callback = EarlyStopping(
+        patience=config.hyper_parameters.patience, monitor="val_loss", verbose=True, mode="min"
+    )
+    print_epoch_result_callback = PrintEpochResultCallback("train", "val")
+    lr_logger = LearningRateMonitor("step")
+
+    gpu = 1 if torch.cuda.is_available() else None
+    trainer = Trainer(
+        max_epochs=config.hyper_parameters.n_epochs,
+        gradient_clip_val=config.hyper_parameters.clip_norm,
+        deterministic=True,
+        check_val_every_n_epoch=config.val_every_epoch,
+        log_every_n_steps=config.log_every_epoch,
+        gpus=gpu,
+        progress_bar_refresh_rate=config.progress_bar_refresh_rate,
+        callbacks=[
+            checkpoint_callback,
+            lr_logger,
+            early_stopping_callback,
+            print_epoch_result_callback,
+        ],
+        resume_from_checkpoint=model_path,
+    )
+
+    trainer.fit(model=model, datamodule=data_module)
+    trainer.test()
+
+    return checkpoint_callback.best_model_path
+
+
+def fine_tune(dataset_path: str, model_path: str, folds_number: int):
+    """Do k-fold cross-validation and compare quality metrics before and after fine-tuning"""
+
+    project_name = os.path.basename(os.path.normpath(dataset_path))
+    start_path = os.path.join(dataset_path, "java-med-psi-no-types")
+    dataset = open(os.path.join(start_path, "java-med-psi-no-types.test.c2s"), "r")
+    samples = dataset.readlines()
+    fold_size = len(samples) // folds_number
```

Comment:
The result rounds to a wrong number (floor but most likely you need ceil)

Reply:
It can cause out-of-bounds on  samples  list